clients/src/main/java/org/apache/kafka/common/utils/Checksums.java

```diff
@@ -40,10 +55,32 @@ public static void update(Checksum checksum, ByteBuffer buffer, int length) {
     public static void update(Checksum checksum, ByteBuffer buffer, int offset, int length) {
         if (buffer.hasArray()) {
             checksum.update(buffer.array(), buffer.position() + buffer.arrayOffset() + offset, length);
-        } else {
-            int start = buffer.position() + offset;
-            for (int i = start; i < start + length; i++)
-                checksum.update(buffer.get(i));
+            return;
+        }
+        if (BYTE_BUFFER_UPDATE != null) {
+            final int oldPosition = buffer.position();
+            final int oldLimit = buffer.limit();
+            try {
+                // save a slice to be used to save an allocation in the hot-path
+                final int start = buffer.position() + offset;
+                buffer.position(0);
+                buffer.limit(start + length);
+                buffer.position(start);
+                BYTE_BUFFER_UPDATE.invokeExact(checksum, buffer);
+                return;
+            } catch (Throwable t) {
+                // fallback
```

Comment:
Shouldn't we propagate the exception instead of falling back?

Reply:
The problem is the existing `update` signature: `MethodHandle` doesn't know about the `update` methods and it just assume that any method can raise a generic `Throwable`, even if it won't happen (as it's in this case).
If I leave the `Throwable` to leak, given that's a checked exception, it's going to poison the existing code using it: I can wrap it in a generic `IllegalStateException` and do it